pennylane/pauli/conversion.py

```diff
@@ -220,7 +129,212 @@ def pauli_decompose(
 
     coeffs = qml.math.stack(coeffs)
 
-    # Convert to Hamiltonian and PauliSentence
+    if not pauli:
+        with qml.QueuingManager.stop_recording():
+            obs = [reduce(matmul, [op_map[o](w) for o, w in obs_term]) for obs_term in obs]
+
+    return (coeffs, obs)
+
+
+def pauli_decompose(
+    H, hide_identity=False, wire_order=None, pauli=False, check_hermitian=True
+) -> Union[Hamiltonian, PauliSentence]:
+    r"""Decomposes a Hermitian matrix into a linear combination of Pauli operators.
+
+    Args:
+        H (array[complex]): a Hermitian matrix of dimension :math:`2^n\times 2^n`.
+        hide_identity (bool): does not include the Identity observable within
+            the tensor products of the decomposition if ``True``.
+        wire_order (list[Union[int, str]]): the ordered list of wires with respect
+            to which the operator is represented as a matrix.
+        pauli (bool): return a PauliSentence instance if ``True``.
+        check_hermitian (bool): check if the provided matrix is Hermitian if ``True``.
+
+    Returns:
+        Union[~.Hamiltonian, ~.PauliSentence]: the matrix decomposed as a linear combination
+        of Pauli operators, either as a :class:`~.Hamiltonian` or :class:`~.PauliSentence` instance.
+
+    .. seealso::
+
+        :func:`~.pauli_decompose_with_phase` for decomposing any general matrix.
+
+    **Example:**
+
+    We can use this function to compute the Pauli operator decomposition of an arbitrary Hermitian
+    matrix:
+
+    >>> A = np.array(
+    ... [[-2, -2+1j, -2, -2], [-2-1j,  0,  0, -1], [-2,  0, -2, -1], [-2, -1, -1,  0]])
+    >>> H = qml.pauli_decompose(A)
+    >>> print(H)
+    (-1.5) [I0 X1]
+    + (-1.5) [X0 I1]
+    + (-1.0) [I0 I1]
+    + (-1.0) [I0 Z1]
+    + (-1.0) [X0 X1]
+    + (-0.5) [I0 Y1]
+    + (-0.5) [X0 Z1]
+    + (-0.5) [Z0 X1]
+    + (-0.5) [Z0 Y1]
+    + (1.0) [Y0 Y1]
+
+    We can return a :class:`~.PauliSentence` instance by using the keyword argument ``pauli=True``:
+
+    >>> ps = qml.pauli_decompose(A, pauli=True)
+    >>> print(ps)
+    -1.0 * I
+    + -1.5 * X(1)
+    + -0.5 * Y(1)
+    + -1.0 * Z(1)
+    + -1.5 * X(0)
+    + -1.0 * X(0) @ X(1)
+    + -0.5 * X(0) @ Z(1)
+    + 1.0 * Y(0) @ Y(1)
+    + -0.5 * Z(0) @ X(1)
+    + -0.5 * Z(0) @ Y(1)
+
+    We can also set custom wires using the ``wire_order`` argument:
+
+    >>> ps = qml.pauli_decompose(A, pauli=True, wire_order=['a', 'b'])
+    >>> print(ps)
+    -1.0 * I
+    + -1.5 * X(b)
+    + -0.5 * Y(b)
+    + -1.0 * Z(b)
+    + -1.5 * X(a)
+    + -1.0 * X(a) @ X(b)
+    + -0.5 * X(a) @ Z(b)
+    + 1.0 * Y(a) @ Y(b)
+    + -0.5 * Z(a) @ X(b)
+    + -0.5 * Z(a) @ Y(b)
+    """
+    n = int(qml.math.log2(len(H)))
+    N = 2**n
+
+    if H.shape != (N, N):
+        raise ValueError("The matrix should have shape (2**n, 2**n), for any qubit number n>=1")
+
+    if check_hermitian and not qml.math.allclose(H, qml.math.conjugate(qml.math.transpose(H))):
```

Comment:
might it be cheaper to just check if any coeffs have an imaginary component after computing it? this is obviously faster if it _isn't_ Hermitian, but hopefully users tend to give Hermitian matrices. As I type this I feel like I'm trying to optimize a tiny part of this so I guess it doesn't matter but just putting it out there.

Reply:
I just borrowed this check from the previous implementation for `pauli_decompose` and added a keyword argument to skip this check. Even I was inclined towards adding the check you mentioned instead of this.